src/libs/ajax.js

```diff
@@ -58,14 +58,60 @@ const withErrorRejection = wrappedFetch => async (...args) => {
   }
 }
 
+const withUrlPrefix = prefix => wrappedFetch => async (path, ...args) => {
+  return wrappedFetch(prefix + path, ...args)
+}
+
+const checkRequesterPaysError = async response => {
+  if (response.status === 400) {
+    const data = await response.json()
+    const requesterPaysError = _.includes('requester pays', _.get(['error', 'message'], data))
+    return Object.assign(new Response(new Blob([JSON.stringify(data)]), response), { requesterPaysError })
+  } else {
+    return Object.assign(response, { requesterPaysError: false })
+  }
+}
+
+const mergeQueryParams = (params, urlString) => {
+  const url = new URL(urlString)
+  url.search = qs.stringify({ ...qs.parse(url.search, { ignoreQueryPrefix: true, plainObjects: true }), ...params })
+  return url.href
+}
+
+/*
+ * Detects errors due to requester pays buckets, and adds the current workspace's billing
+ * project if the user has access, retrying the request once if necessary.
+ */
+const withRequesterPays = wrappedFetch => async (url, ...args) => {
+  const bucket = /\/b\/([^/]+)\//.exec(url)[1]
+  const workspace = workspaceStore.get()
+  const userProject = workspace && Utils.canWrite(workspace.accessLevel) && workspace.workspace.namespace
+  const tryRequest = async () => {
+    const knownRequesterPays = _.includes(bucket, requesterPaysBuckets.get())
+    try {
+      return await wrappedFetch(mergeQueryParams({ userProject: (knownRequesterPays && userProject) || undefined }, url), ...args)
+    } catch (error) {
+      const newResponse = await checkRequesterPaysError(error)
+      if (newResponse.requesterPaysError && !knownRequesterPays) {
+        requesterPaysBuckets.update(_.union([bucket]))
+        if (userProject) {
+          return tryRequest()
+        }
+      }
+      throw newResponse
+    }
+  }
+  return tryRequest()
+}
+
 const fetchOk = _.flow(withInstrumentation, withCancellation, withErrorRejection)(fetch)
 
 
 const fetchSam = async (path, options) => {
   return fetchOk(`${getConfig().samUrlRoot}/${path}`, addAppIdentifier(options))
 }
 
-const fetchBuckets = (path, options) => fetchOk(`https://www.googleapis.com/${path}`, options)
+const fetchBuckets = _.flow(withRequesterPays, withUrlPrefix('https://www.googleapis.com/'))(fetchOk)
```

Comment:
`withUrlPrefix` feels a bit indirect and specific to this case. It is very dependent on the shape of the args of the function that the flow is targeting. Would it be possible to generalize this to something for transforming function arguments so that calling would look more like:
```
const fetchBuckets = _.flow(
  withRequesterPays,
  withArgTransform({ path, ...options } => { `https://www.googleapis.com/${path}`, ...options }
)(fetchOk)
```
This may or may not be a workable implementation (might not even compile):
```
const withArgTransform transform => wrapped => args => wrapped(transform(args))
```
I'm not in love with the name `withArgTransform`. It's a little long for my liking. Other thoughts?

Reply:
This is specifically meant to be a 'fetch transformer', like the other 'with-' functions in this file, so it does expect a particular signature. Adding a prefix seems the be an overwhelmingly common case, so it seemed like a decent abstraction. (I didn't convert other places so as not to clutter up this PR)

Reply:
Ahh, so naming the parameter `wrappedFetch` is very intentional. I get it now. (I am making progress with internalizing these functional patterns but not quite there yet, so things like this look more opaque to me.)